blacklists.py

```diff
@@ -140,17 +147,174 @@ def exists(self, item: Union[str, dict]):
+        for item in y['items']:
+            if not keep_disabled and 'disable' in item and item['disable']:
+                continue
+            yield item
+
+    def parse(self):
+        for item in self._parse():
+            yield item[self.__class__.SCHEMA_PRIKEY]
+
+    def _write(self, callback):
+        d = {
+            'Schema': self.__class__.SCHEMA_VARIANT,
+            'Schema_version': self.__class__.SCHEMA_VERSION,
+            'items': sorted(
+                self._parse(keep_disabled=True),
+                key=lambda x: x[self.__class__.SCHEMA_PRIKEY])
+        }
+        callback(d)
+        with open(self._filename, 'w', encoding='utf-8') as f:
+            yaml.dump(d, f)
+
+    def _validate(self, item):
+        ip_regex = regex.compile(
+            r'(?(DEFINE)(?P<octet>'
+            '' r'1[0-9]{0,2}|2(?:[0-4][0-9])??|25[0-5]?|2[6-9]|[3-9][0-9]?))'
+            r'^(?&octet)(?:\.(?&octet)){3}$'
+        )
+        if 'ip' in item:
+            if not ip_regex.match(item['ip']):
+                raise ValueError('Field "ip" is not a valid IP address: {0}'.format(
+                    item['ip']))
+            '''
+            if 'cidr' in item:
+                raise ValueError(
+                    'Cannot have both "ip" and "cidr" members: {0!r}'.format(item))
+        elif 'cidr' in item:
+            if not 'base' in item['cidr'] or not 'mask' in item['cidr']:
+                raise ValueError('Field "cidr" must have members "base" and "mask"')
+            if not ip_regex.match(item['cidr']['base']):
+                raise ValueError('Field "base" is not a valid IP address: {0}'.format(
+                    item['cidr']['base']))
+            mask = int(item['cidr']['mask'])
+            if mask < 0 or mask > 32:
+                raise ValueError('Field "mask" must be between 0 and 32: {0}'.format(
+                    item['cidr']['mask']))
+            '''
+        else:
+            raise ValueError('Item needs to have an "ip" member field: {0!r}'.format(item))
+
+    def add(self, item):
+        self._validate(item)
+        prikey = self.__class__.SCHEMA_PRIKEY
+
+        def add_callback(d):
+            for compare in d['items']:
+                if compare[prikey] == item[prikey]:
+                    raise ValueError('{0} already in list {1}'.format(
+                        item[prikey], d['items']))
+            d['items'].append(item)
 
+        self._write(add_callback)
+
+    def remove(self, item):
+        prikey = self.__class__.SCHEMA_PRIKEY
+
+        def remove_callback(d):
+            for i, compare in enumerate(d['items']):
+                if compare[prikey] == item[prikey]:
+                    break
+            else:
+                raise ValueError('No {0} found in list {1}'.format(
+                    item[prikey], d['items']))
+            del d['items'][i]
+
+        self._write(remove_callback)
+
+    # FIXME: enumerate gets YAML item array index, not line number
```

Comment:
I'm afraid this isn't easily solvable. We could just stick with YAML array index and rely on Ctrl-F (or Vim `/`) for quick lookup.

Reply:
I found a Stack Overflow recipe for this but let's tackle that separately if that seems important. I guess the `!!/bisect` functionality is where we need to display the line number; perhaps that could be refactored to not always assume the result is a line number.

Reply:
Here's a Stack Overflow question with a couple of suggested answers which look at least vaguely promising: https://stackoverflow.com/questions/13319067/parsing-yaml-return-with-line-number